Sources/AsyncAlgorithms/Debounce/DebounceStorage.swift

```diff
@@ -0,0 +1,296 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Async Algorithms open source project
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+final class DebounceStorage<Base: AsyncSequence, C: Clock>: @unchecked Sendable where Base: Sendable {
+    typealias Element = Base.Element
+
+    /// The state machine protected with a lock.
+    private let stateMachine: ManagedCriticalState<DebounceStateMachine<Base, C>>
+    /// The interval to debounce.
+    private let interval: C.Instant.Duration
+    /// The tolerance for the clock.
+    private let tolerance: C.Instant.Duration?
+    /// The clock.
+    private let clock: C
+
+    init(base: Base, interval: C.Instant.Duration, tolerance: C.Instant.Duration?, clock: C) {
+        self.stateMachine = .init(.init(base: base, clock: clock, interval: interval))
+        self.interval = interval
+        self.tolerance = tolerance
+        self.clock = clock
+    }
+
+    func iteratorDeinitialized() {
+        let action = self.stateMachine.withCriticalRegion { $0.iteratorDeinitialized() }
+
+        switch action {
+        case .cancelTaskAndUpstreamAndClockContinuations(
+            let task,
+            let upstreamContinuation,
+            let clockContinuation
+        ):
+            upstreamContinuation?.resume(throwing: CancellationError())
+            clockContinuation?.resume(throwing: CancellationError())
+
+            task.cancel()
+
+        case .none:
+            break
+        }
+    }
+
+    func next() async rethrows -> Element? {
+        // We need to handle cancellation here because we are creating a continuation
+        // and because we need to cancel the `Task` we created to consume the upstream
+        return try await withTaskCancellationHandler {
+            // We always suspend since we can never return an element right away
+
+            let result: Result<Element?, Error> = await withUnsafeContinuation { continuation in
+                self.stateMachine.withCriticalRegion {
+                    let action = $0.next(for: continuation)
+
+                    switch action {
+                    case .startTask(let base):
+                        self.startTask(
```

Comment:
This is also safe to do since we are starting the `Task` while holding the lock. If we are in a single threaded env the new `Task` will get created but it will never make any progress until the current job (which is calling `next()`) is finished.

If we are not holding the lock across the start of the `Task` we get weird race conditions where the upstream consumption child tasks could call out to the state machine before we even transitioned from `initial` which I want to avoid and would cause unnecessary complexity in the state handling.